File: src/fairseq2/data/processors.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

from typing import TYPE_CHECKING, List, Optional, Sequence

from torch import Tensor

from fairseq2 import _DOC_MODE
from fairseq2.data.typing import StringLike
from fairseq2.typing import DataType

if TYPE_CHECKING or _DOC_MODE:

    class StrSplitter:
        def __init__(self, sep: str = "\t") -> None:
            ...

        def __call__(self, s: StringLike) -> List[StringLike]:
            ...

    class StrToTensorConverter:
        def __init__(
            self,
            size: Optional[Sequence[int]] = None,
            dtype: Optional[DataType] = None,
        ) -> None:
            ...

        def __call__(self, s: StringLike) -> Tensor:
            ...

else:
    from fairseq2.C.data.processors import StrSplitter as StrSplitter
    from fairseq2.C.data.processors import StrToTensorConverter as StrToTensorConverter

    def _set_module_name() -> None:
        for t in [StrSplitter, StrToTensorConverter]:
            t.__module__ = __name__

    _set_module_name()

File: src/fairseq2/native/extension/data/processors.cc
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "fairseq2/native/extension/module.h"

#include <cstdint>
#include <optional>
#include <stdexcept>
#include <string_view>
#include <utility>
#include <vector>

#include <ATen/ScalarType.h>

#include <fairseq2/native/data/data_pipeline.h>
#include <fairseq2/native/data/data_processor.h>
#include <fairseq2/native/data/processors/string_splitter.h>
#include <fairseq2/native/data/processors/string_to_integer_converter.h>
#include <fairseq2/native/data/processors/string_to_tensor_converter.h>

namespace py = pybind11;

namespace fairseq2 {

void
def_processors(py::module_ &data_module)
{
    py::module_ m = data_module.def_submodule("processors");

    // StrSplitter
    py::class_<string_splitter, data_processor, std::shared_ptr<string_splitter>>(m, "StrSplitter")
        .def(
            py::init([](std::string_view sep, std::optional<std::vector<std::string>> names)
            {
                if (sep.size() != 1)
                    throw std::invalid_argument{"`sep` must be of length 1."};

                return string_splitter{sep[0], std::move(names)};
            }),
            py::arg("sep") = '\t',
            py::arg("names") = std::nullopt);

    // StrToIntConverter
    py::class_<
        string_to_integer_converter,
        data_processor,
        std::shared_ptr<string_to_integer_converter>>(m, "StrToIntConverter")
        .def(py::init<std::int16_t>(), py::arg("base") = 10);

    // StrToTensorConverter
    py::class_<
        string_to_tensor_converter,
        data_processor,
        std::shared_ptr<string_to_tensor_converter>>(m, "StrToTensorConverter")
        .def(
            py::init<std::optional<std::vector<std::int64_t>>, std::optional<at::ScalarType>>(),
            py::arg("size") = std::nullopt,
            py::arg("dtype") = std::nullopt);
}

}  // namespace fairseq2
